## Курсова робота на тему "Визначення найбільш вигідної роздрібної ціни продажу товару в магазині з урахуванням плати на кредит"
## Виконав студент гр. ІС-72 Цилюрик Дмитро

### Програмний опис імітації мережі Петрі використовуючи ООП підхід

##### Імпорт необхідних бібліотек

In [17]:
import json
import math
import numpy as np
import pandas as pd
import sys
from enum import Enum
from IPython.display import display

##### Встановлення неохідних опцій для подальшого використання

In [18]:
pd.set_option('display.max_rows', 5000)
pd.set_option('display.max_columns', 5000)
pd.set_option('display.width', 1000)
pd.set_option('mode.chained_assignment', None)
pd.set_option('display.float_format', lambda x: '%.5f' % x)

##### Клас збереження типів розподілу для ренерації затримок

In [2]:
class Distribution(Enum):
    EXPONENTIAL = np.random.exponential
    UNIFORM = np.random.uniform
    NORMAL = np.random.normal
    POISSON = np.random.poisson

##### Клас для генерації затримок

In [3]:
class RandGenerator:
    @staticmethod
    def generate(distribution_function, **args):
        return distribution_function(**args)

##### Клас для представлення Позиції в мережі Петрі

In [4]:
class Position:
    def __init__(self, num_of_markers=0, description=None, **kwargs):
        self.num_of_markers = num_of_markers
        self.description = description

        self.in_arcs = []
        self.out_arcs = []

    def add_markers(self, num_of_markers):
        self.num_of_markers += num_of_markers

    def remove_markers(self, num_of_markers):
        self.num_of_markers -= num_of_markers

    def __repr__(self):
        return f"Position<{self.description}> = {self.num_of_markers}"

##### Клас для представлення Переходу в мережі Перті

In [5]:
class Transition:
    def __init__(
        self,
        delay=0,
        delay_distribution=None,
        delay_distribution_params=None,
        priority=1,
        probability=None,
        description=None,
        **kwargs,
    ):
        self.delay = delay
        self.delay_distribution = delay_distribution
        self.delay_distribution_params = delay_distribution_params
        self.priority = priority
        self._probability = probability
        self.description = description

        self.in_arcs = []
        self.out_arcs = []

        self.markers_release_times = {}

    @property
    def probability(self):
        return self._probability
    
    def _update_markers_release_times(self, release_timestamp):
        if release_timestamp in self.markers_release_times:
            self.markers_release_times[release_timestamp] += 1
        else:
            self.markers_release_times[release_timestamp] = 1
    
    def get_delay(self):
        return round(
            self.delay 
            if self.delay_distribution is None 
            else RandGenerator.generate(
                self.delay_distribution.value, **self.delay_distribution_params
            )
        )

    def check_can_perform_move_from_transition(self):
        return all(arc.check_move_possibility() for arc in self.in_arcs)

    def perform_move_from_transition(self, current_time):
        delay = self.get_delay()
        release_timestamp = current_time + delay
        self._update_markers_release_times(release_timestamp)
        for arc in self.in_arcs:
            arc.move_from()

    def check_can_perform_move_to_transition(self, current_time):
        return current_time in self.markers_release_times.keys()
    
    def perform_move_to_transition(self, current_time):
        for _ in range(self.markers_release_times[current_time]):
            for arc in self.out_arcs:
                arc.move_to()
        else:
            del self.markers_release_times[current_time]

    def __repr__(self):
        return f"Transition<{self.description}> = {self.markers_release_times}"

##### Клас для представлення Переходу в мережі Перті із заданою функцієї умовірностей переходу

In [20]:
class TransitionWithProbabilityFunc(Transition):
    def __init__(self, probability_func, **kwargs):
        super().__init__(**kwargs)
        self._queue = kwargs["queue"]
        self._m = kwargs["m"]
        self._probability_func = probability_func

    @property
    def probability(self):
        return self._probability_func(self._queue.num_of_markers, self._m)

##### Клас для представлення Дуги в мережі Перті

In [21]:
class Arc:
    def __init__(self, start, end, multiplier=1, informational=False):
        self.start = start
        self.end = end
        self.multiplier = multiplier
        self.informational = informational

        self._set_arc_ends()

    def _set_arc_ends(self):
        self.start.out_arcs.append(self)
        self.end.in_arcs.append(self)

    def check_move_possibility(self):
        return self.start.num_of_markers >= self.multiplier

    def move_from(self):
        if not self.informational:
            self.start.remove_markers(self.multiplier)

    def move_to(self):
        self.end.add_markers(self.multiplier)

    def __repr__(self):
        return f"{self.start} --> {self.end}"

##### Клас для представлення Дуги в мережі Перті за умови переходу всіх мартерів із Позиції

In [22]:
class ArcAll(Arc):
    def __init__(self, move_to_all=True, **kwargs):
        super().__init__(**kwargs)
        self._move_to_all = move_to_all

    def check_move_possibility(self):
        return self.start.num_of_markers >= 0

    def move_from(self):
        self.end.multiplier_all = self.start.num_of_markers
        if not self.informational:
            self.start.remove_markers(self.start.num_of_markers)

    def move_to(self):
        if self._move_to_all:
            self.end.add_markers(self.start.multiplier_all)
        else:
            super().move_to()

##### Клас для представлення Дуги в мережі Перті за умови переходу всіх мартерів із Позиції за заданої функцію визначення кількості маркерів для переходу

In [23]:
class ArcAllWithMoveToFunc(ArcAll):
    def __init__(self, move_to_func, **kwargs):
        super().__init__(**kwargs)
        self._move_to_func = move_to_func
    
    def move_to(self):
        if self._move_to_all:
            self.end.add_markers(self._move_to_func(self.start.multiplier_all))
        else:
            super().move_to()

##### Клас моделювання мережі Петрі

In [186]:
class Model:
    def __init__(
        self, 
        transitions, 
        positions, 
        modeling_time,
        stop_condition=None,
        verbose=False,
    ):
        self.transitions = transitions
        self.positions = positions
        self._modeling_time = modeling_time
        self._verbose = verbose
        
        if stop_condition is None:
            self._stop_condition = lambda : True
        else:
            self._stop_condition = stop_condition
        
        self.current_time = 0
        self.position_markers_stats = {position: [] for position in positions}

    def run(self):
        if self._verbose:
            self._display_model_status()
        while self.current_time < self._modeling_time:
            while True:       
                if not any(
                    transition.check_can_perform_move_from_transition()
                    for transition in self.transitions
                ):
                    break
                self._move()
                self.update_statistics()
                
            if not any(
                len(transition.markers_release_times) > 0
                for transition in self.transitions
            ) or self._stop_condition():
                break
            self._move()
            self.update_statistics()
            
            self.current_time += 1
            if self._verbose:
                self._display_model_status()

        return self.get_result_statistics()

    def _move(self):
        transitions_without_conflicts = self._get_transitions_with_resolved_conflicts()
        for transition in transitions_without_conflicts:
            if transition.check_can_perform_move_from_transition():
                transition.perform_move_from_transition(self.current_time)
        for transition in self.transitions:
            if transition.check_can_perform_move_to_transition(self.current_time):
                transition.perform_move_to_transition(self.current_time)
  
    def _get_transitions_with_resolved_conflicts(self):
        conflicting_transitions = []
        resulting_transitions = []

        for position in self.positions:
            non_informational_arcs = filter(lambda arc: not arc.informational, position.out_arcs)

            valid_transitions = [*filter(
                lambda transition: transition.check_can_perform_move_from_transition(),
                map(lambda arc: arc.end, non_informational_arcs)
            )]

            if len(valid_transitions) > 1:
                conflicting_transitions.append(valid_transitions)

        for conflict in conflicting_transitions:
            is_probabilistic_conflict = conflict[0].probability is not None

            if is_probabilistic_conflict:
                probabilities = [*map(lambda transition: transition.probability, conflict)]

                gap = 1 - sum(probabilities)
                if gap != 0:

                    probabilities = [p + gap / len(probabilities) for p in probabilities]

                resulting_transitions.append(np.random.choice(conflict, p=probabilities))
            else:
                conflict.sort(key=lambda transition: transition.priority, reverse=True)
                resulting_transitions.append(conflict[0])


        for transition in self.transitions:
            was_transition_in_conflict = any([transition in conflict for conflict in conflicting_transitions])

            if not was_transition_in_conflict:
                resulting_transitions.append(transition)

        return resulting_transitions

    def update_statistics(self):
        for position in self.positions:
            self.position_markers_stats[position].append(position.num_of_markers)

    def get_result_statistics(self):
        result_stats = []

        for position, stats in self.position_markers_stats.items():
            result_stats.append(
                {
                    "description": position.description,
                    "avg_markers": sum(stats) / len(stats),
                    "max_markers": max(stats),
                    "min_markers": min(stats),
                    "result_markers": stats[-1],
                    "time_modeling": self.current_time,
                }            
            )

        return result_stats

    def _display_model_status(self):
        print(f"Current time: {self.current_time}")
        for position in self.positions:
            print(position)
        for transition in self.transitions:
            print(transition)
        print()

![Model](diagrams/petri_model.png "Мережа Петрі для задачі")

### Моделювання задачі

In [201]:
class StoreSystem:
    LAMBDA = 30
    k = 0.02
    s = 500
    N = 1000

    p1 = {
        (0, 5): 0.25,
        (6, 10): 0.25,
        (11, 20): 0.2,
        (21, 50): 0.2,
        (51, sys.maxsize): 0.1,
    }
    p2 = {
        (0, 1.4999): 0.3,
        (1.5, 1.9999): 0.3,
        (2.0, 2.9999): 0.2,
        (3.0, 3.9999): 0.1,
        (4.0, 5.9999): 0.06,
        (6.0, 9.9999): 0.04,
        (10.0, sys.maxsize): 0,
    }

    A = 30
    B = 5

    M = 1000
    minutes_in_one_day = 24 * 60
    modeling_time = math.inf * minutes_in_one_day
       
    def __init__(self, m):
        self.m = m
        
        # Positions
        self.customer_generator = Position(num_of_markers=1, description="Потік покупців")
        self.customer_came = Position(num_of_markers=0, description="Покупець надійшов")
        self.queue = Position(num_of_markers=0, description="Черга покупців")
        self.customer_left_store = Position(num_of_markers=0, description="Кількість покупців, що покинули мазагин")
        self.stock = Position(num_of_markers=self.N, description="Запас товару")
        self.seller = Position(num_of_markers=1, description="Продавець")
        self.customer_success = Position(num_of_markers=0, description="Кількість обслугованих покупців")
        self.profit = Position(num_of_markers=0, description="Прибуток")
        self.debt = Position(num_of_markers=self.N * self.s * self.M, description="Борг")
        self.repaid_debt = Position(num_of_markers=0, description="Погашений борг")
        self.delay_generator = Position(num_of_markers=1, description="Затримка процентної ставки")
        self.interest_rate_ready = Position(num_of_markers=0, description="Процентна ставка готова до дії")
        
        # Transitions
        self.customer_arrive = Transition(
            delay_distribution=Distribution.EXPONENTIAL,
            delay_distribution_params={"scale": self.LAMBDA},
            description="Надходження покупця"
        )
        self.customer_leave_store = TransitionWithProbabilityFunc(
            probability_func=self.get_leave_probability, 
            queue=self.queue,
            m=self.m,
            description="Покупець покипає магазин"
        )
        self.customer_stay_in_store = TransitionWithProbabilityFunc(
            probability_func=self.get_stay_in_store_probability, 
            queue=self.queue,
            m=self.m,
            description="Покупець стає в чергу"
        )
        self.service = Transition(
            delay_distribution=Distribution.UNIFORM,
            delay_distribution_params={"low": self.A - self.B, "high": self.A + self.B},
            description="Обслуговування покупця"
        )
        self.debt_repayment = Transition(description="Гасіння боргу")
        self.interest_rate_delay = Transition(delay=self.minutes_in_one_day, description="Затримка для процентнох ставки")
        self.interest_rate_effect = Transition(description="Дія процентної ставки", priority=2)
        
        # Arcs
        Arc(start=self.customer_generator, end=self.customer_arrive)
        Arc(start=self.customer_arrive, end=self.customer_generator)
        Arc(start=self.customer_arrive, end=self.customer_came)
        Arc(start=self.customer_came, end=self.customer_leave_store)
        Arc(start=self.customer_leave_store, end=self.customer_left_store)
        Arc(start=self.customer_came, end=self.customer_stay_in_store)
        Arc(start=self.customer_stay_in_store, end=self.queue)
        Arc(start=self.queue, end=self.service)
        Arc(start=self.stock, end=self.service)
        Arc(start=self.seller, end=self.service)
        Arc(start=self.service, end=self.seller)
        Arc(start=self.service, end=self.customer_success)
        Arc(start=self.service, end=self.profit, multiplier=int(self.m * self.s * self.M))
        Arc(start=self.profit, end=self.debt_repayment, multiplier=self.M)
        Arc(start=self.debt, end=self.debt_repayment, multiplier=self.M)
        Arc(start=self.debt_repayment, end=self.repaid_debt, multiplier=self.M)
        Arc(start=self.delay_generator, end=self.interest_rate_delay)
        Arc(start=self.interest_rate_delay, end=self.delay_generator)
        Arc(start=self.interest_rate_delay, end=self.interest_rate_ready)
        Arc(start=self.interest_rate_ready, end=self.interest_rate_effect)
        ArcAll(start=self.debt, end=self.interest_rate_effect)
        ArcAllWithMoveToFunc(start=self.interest_rate_effect, end=self.debt, move_to_func=self.calculate_debt_rate_effect)


    @staticmethod
    def get_stay_in_store_probability(num_in_queue, m):
        p1_current = next(filter(lambda item: item[0][0] <= num_in_queue <= item[0][1], StoreSystem.p1.items()))[1]
        p2_current = next(filter(lambda item: item[0][0] <= m <= item[0][1], StoreSystem.p2.items()))[1]
        return p1_current * p2_current

    @staticmethod
    def get_leave_probability(num_in_queue, m):
        return 1 - StoreSystem.get_stay_in_store_probability(num_in_queue, m)

    @staticmethod
    def calculate_debt_rate_effect(current_debt):
        return round(current_debt * (1 + StoreSystem.k))
    
    def _stop_condition(self):
        return (
            True 
            if self.stock.num_of_markers == 0 and len(self.service.markers_release_times) == 0
            else False
        )
    
    def _append_target_function_result(self, modeling_results):
        modeling_results.append(
            {
                "description": "Чистий прибуток",
                "avg_markers": None,
                "max_markers": None,
                "min_markers": None,
                "result_markers": (self.profit.num_of_markers - self.debt.num_of_markers) / StoreSystem.M,
                "time_modeling": modeling_results[0]["time_modeling"],
            }
        )
        return modeling_results

    def simulate(self, verbose=False, flag_display_results=True):
        net = Model(
            positions=[
                self.customer_generator,
                self.customer_came,
                self.queue,
                self.customer_left_store,
                self.stock,
                self.seller,
                self.customer_success,
                self.profit,
                self.debt,
                self.repaid_debt,
                self.delay_generator,
                self.interest_rate_ready,
            ],
            transitions=[
                self.customer_arrive,
                self.customer_leave_store,
                self.customer_stay_in_store,
                self.service,
                self.debt_repayment,
                self.interest_rate_delay,
                self.interest_rate_effect,
            ],
            modeling_time=self.modeling_time,
            stop_condition=self._stop_condition,
            verbose=verbose,
        )
        modeling_results = self._append_target_function_result(net.run())
        
        if flag_display_results:
            display(pd.DataFrame(modeling_results))
        
        return modeling_results


### Оцінка адекватності моделі

In [202]:
_system = StoreSystem(1.7)
test_simulation = _system.simulate()

description   avg_markers   max_markers  min_markers  result_markers  time_modeling
0                            Потік покупців  1.037946e-02  1.000000e+00          0.0    0.000000e+00         411539
1                         Покупець надійшов  1.037946e-02  1.000000e+00          0.0    0.000000e+00         411539
2                            Черга покупців  2.298013e-03  1.000000e+00          0.0    1.000000e+00         411539
3   Кількість покупців, що покинули мазагин  6.181312e+03  1.224200e+04          0.0    1.224200e+04         411539
4                              Запас товару  5.019014e+02  1.000000e+03          0.0    0.000000e+00         411539
5                                 Продавець  9.249721e-01  1.000000e+00          0.0    1.000000e+00         411539
6           Кількість обслугованих покупців  4.980236e+02  1.000000e+03          0.0    1.000000e+03         411539
7                                  Прибуток  2.834712e+05  8.500000e+05          0.0    0.000000e+00         411539
8                                      Борг  1.760730e+10  9.788040e+10  493200000.0    9.787870e+10         411539
9                            Погашений борг  4.230366e+08  8.500000e+08          0.0    8.500000e+08         411539
10               Затримка процентної ставки  2.233744e-04  1.000000e+00          0.0    0.000000e+00         411539
11           Процентна ставка готова до дії  2.233744e-04  1.000000e+00          0.0    0.000000e+00         411539
12                          Чистий прибуток           NaN           NaN          NaN   -9.787870e+07         411539

#### Середній час надходження покупців

In [224]:
num_of_clients = 0
for item in test_simulation:
    if item["description"] in ["Кількість покупців, що покинули мазагин", "Кількість обслугованих покупців"]:
        num_of_clients += item["result_markers"]

print(f"Average arrival time: {test_simulation[0]['time_modeling']/num_of_clients}")

31.07831143331823

Як можна помітити, середній час надходження покупців є близьким до дійсності. За умовою клієнти надходитя в чередньому кожні 30 хлилин (пуасонівський потік із параметром λ = 1/30 ($\frac{1}{хв}$)

#### Ймовірність купівлі товару

In [230]:
failed_num_of_clients, succeeded_num_of_clients = 0, 0
for item in test_simulation:
    if item["description"] == "Кількість покупців, що покинули мазагин":
        failed_num_of_clients += item["result_markers"]
    elif item["description"] == "Кількість обслугованих покупців":
        succeeded_num_of_clients += item["result_markers"]

print(f"Average probability of purchase: {succeeded_num_of_clients/(succeeded_num_of_clients+failed_num_of_clients)}")

Average probability of purchase: 0.07551729346020239


При запуску тестового прогону був використаний коефіцієнт ціни товару: $m = 1.7$. Це відповідає значенню $p_2=0.3$. Середнє значення покупців в черзі рівне 0 (див. середнє значення в позиції "Черга покупців"), тому $p_1=0.25$. Отримаємо наступну ймовірність купівлі товару покупцем: 
\begin{equation*}
P = p_1*p_2 = 0.3*0.25 = 0.075
\end{equation*}
Теоретична та експериментальна частота покупок є майже однаковими. Отже, модель є адекватною.

### Визначення кількості прогонів
Надалі будемо використовувати дану формулу:
\begin{equation*}
N = \frac{t_α^2σ^2}{ε^2}+1
\end{equation*}
де $σ^2$ - дисперсія відгуку моделі, $ε$ - точність вимірювання, $t_α$ - агрумент функції Лаплача

In [203]:
experiments = []
for i in np.arange(20):
    print(f"Experiment {i}")
    m = 1.7
    store_system = StoreSystem(m)
    for result_item in store_system.simulate(flag_display_results=False): 
        experiments.append(
            {
                "price_coefficient":m,
                **result_item
            }
        )

Experiment 0
Experiment 1
Experiment 2
Experiment 3
Experiment 4
Experiment 5
Experiment 6
Experiment 7
Experiment 8
Experiment 9
Experiment 10
Experiment 11
Experiment 12
Experiment 13
Experiment 14
Experiment 15
Experiment 16
Experiment 17
Experiment 18
Experiment 19


In [205]:
experiments_df = pd.DataFrame(experiments)
profit_df = experiments_df.loc[experiments_df["description"] == "Чистий прибуток"]

In [206]:
display(profit_df)

price_coefficient      description  avg_markers  max_markers  min_markers  result_markers  time_modeling
12                 1.7  Чистий прибуток          NaN          NaN          NaN   -8.025148e+07         396926
25                 1.7  Чистий прибуток          NaN          NaN          NaN   -9.037609e+07         406900
38                 1.7  Чистий прибуток          NaN          NaN          NaN   -1.118512e+08         420524
51                 1.7  Чистий прибуток          NaN          NaN          NaN   -9.347771e+07         409732
64                 1.7  Чистий прибуток          NaN          NaN          NaN   -1.154950e+08         422022
77                 1.7  Чистий прибуток          NaN          NaN          NaN   -1.466659e+08         437640
90                 1.7  Чистий прибуток          NaN          NaN          NaN   -8.525043e+07         403171
103                1.7  Чистий прибуток          NaN          NaN          NaN   -1.138227e+08         420556
116                1.7  Чистий прибуток          NaN          NaN          NaN   -1.196249e+08         427261
129                1.7  Чистий прибуток          NaN          NaN          NaN   -7.109059e+07         388018
142                1.7  Чистий прибуток          NaN          NaN          NaN   -8.398534e+07         404172
155                1.7  Чистий прибуток          NaN          NaN          NaN   -1.148613e+08         420176
168                1.7  Чистий прибуток          NaN          NaN          NaN   -9.807275e+07         411349
181                1.7  Чистий прибуток          NaN          NaN          NaN   -6.954186e+07         386244
194                1.7  Чистий прибуток          NaN          NaN          NaN   -7.435876e+07         391795
207                1.7  Чистий прибуток          NaN          NaN          NaN   -1.007508e+08         411918
220                1.7  Чистий прибуток          NaN          NaN          NaN   -9.102521e+07         407401
233                1.7  Чистий прибуток          NaN          NaN          NaN   -8.094476e+07         400273
246                1.7  Чистий прибуток          NaN          NaN          NaN   -1.062636e+08         417358
259                1.7  Чистий прибуток          NaN          NaN          NaN   -1.587291e+08         441913

In [221]:
needed_number_of_experiments = round(
    (1.96**2*profit_df.std()["result_markers"]**2) / 
    (profit_df.mean()["result_markers"]*0.15)**2 
    + 1
)

In [225]:
print(f"Number of needed experiments: {needed_number_of_experiments}")

Number of needed experiments: 11.0


### Аналіз та оцінка результатів

Для визначення впливу факторів на значення цільової функції було використано дисперсійний аналіз. Основне питання, на яке дає відповідь дисперсійний аналіз впливу фактора, формулюється так: різниця у значеннях відгуку моделі, отриманих при різних значеннях фактора обумовлена випадковістю, чи пояснюється виключно дією фактора?
Надалі будуть використні наступні формули:
- для розрахунку середнього значення цільової функції заданого коефіцієнта $j$
\begin{equation*}
\overline{y_j} = \frac{1}{n}\sum _{i=1}^{n}y_{ij}, 
\end{equation*}
де $n$ - кількість прогонів
- для розрахунку середнього значення цільової функції
\begin{equation*}
\overline{y} = \frac{1}{m}\sum _{j=1}^{m}y_{j}, 
\end{equation*}
де $m$ - кількість коефіцієнтів
- для розрахунку значення різниць між групами коефіцієнтів
\begin{equation*}
d_{факт} = n \sum _{j=1}^{m}(\overline{y_{j}}-\overline{y})^2, 
\end{equation*}
де $n$ - кількість прогонів, $m$ - кількість коефіцієнтів
- для розрахунку значення через групами коефіцієнтів
\begin{equation*}
d_{залиш} = \frac{\sum _{i=1}^{K}\sum _{j=1}^{n_i}(y_{ij}-\overline{y_{i}})^2}{N-K}, 
\end{equation*}
\begin{equation*}
N = m*n,
\end{equation*}
\begin{equation*}
K = m
\end{equation*}
де $n$ - кількість прогонів, $m$ - кількість коефіцієнтів
- для розрахунку значення критерію Фішера
\begin{equation*}
F = \frac{d_{факт}}{d_{залиш}}
\end{equation*}

Виходячи з контексту даної задачі є доцільним обрати **коефіцієнт ціни товару** (**_price_coefficient_**) фактором, при цьому відгуком моделі буде значення цільової функції, тобто "**Чистий прибуток**" ("**_Прибуток_**" - "**_Борг_**")

Для визначення набору параметрів звернемось до умови задачі. Маємо настуний розподіл ймовірностей в залежності від цінового коефіцієнту:
```python
p2 = {
    (0, 1.4999): 0.3,
    (1.5, 1.9999): 0.3,
    (2.0, 2.9999): 0.2,
    (3.0, 3.9999): 0.1,
    (4.0, 5.9999): 0.06,
    (6.0, 9.9999): 0.04,
    (10.0, sys.maxsize): 0,
}
```
Доцільним є вибір коефіцієнтів які знаходяться на граничному переході, тобто для першої ітерації будуть використані наступні значення коефіцієнтів:
```python
m in [1.0, 1.4, 1.9, 2.9, 3.9, 5.9, 9.9]
```

In [236]:
results_first_iteration = []
for m in [1.0, 1.4, 1.9, 2.9, 3.9, 5.9, 9.9]:
    for i in np.arange(needed_number_of_experiments):
        print(f"price_coefficient: {m}; experiment {i}")
        store_system = StoreSystem(m)
        for result_item in store_system.simulate(flag_display_results=False): 
            results_first_iteration.append(
                {
                    "experiment": i,
                    "price_coefficient":m,
                    **result_item
                }
            )

price_coefficient: 1.0; experiment 0.0
price_coefficient: 1.0; experiment 1.0
price_coefficient: 1.0; experiment 2.0
price_coefficient: 1.0; experiment 3.0
price_coefficient: 1.0; experiment 4.0
price_coefficient: 1.0; experiment 5.0
price_coefficient: 1.0; experiment 6.0
price_coefficient: 1.0; experiment 7.0
price_coefficient: 1.0; experiment 8.0
price_coefficient: 1.0; experiment 9.0
price_coefficient: 1.0; experiment 10.0
price_coefficient: 1.4; experiment 0.0
price_coefficient: 1.4; experiment 1.0
price_coefficient: 1.4; experiment 2.0
price_coefficient: 1.4; experiment 3.0
price_coefficient: 1.4; experiment 4.0
price_coefficient: 1.4; experiment 5.0
price_coefficient: 1.4; experiment 6.0
price_coefficient: 1.4; experiment 7.0
price_coefficient: 1.4; experiment 8.0
price_coefficient: 1.4; experiment 9.0
price_coefficient: 1.4; experiment 10.0
price_coefficient: 1.9; experiment 0.0
price_coefficient: 1.9; experiment 1.0
price_coefficient: 1.9; experiment 2.0
price_coefficient: 1.9;

_Для уникнення очікування виконання експериментів, результати були збережені. Для відновлення необхідно запустити наступну ділянку коду_

In [19]:
results_first_iteration_path = (
    "experiments/price_coefficient_change_11_runs_1_iteration.json"
)
results_first_iteration = []
with open(results_first_iteration_path, "r") as f:
    for line in f.readlines():
        results_first_iteration.append(json.loads(line))

Аналіз результатів та перевірка значимості фактора за критерієм Фішера.
**_Помітка_**: для визначння ```F_critical``` був використаний сервіс [![Free Statistics Calculators](references/free_statistics_calculators.png "Free Statistics Calculators")](https://www.danielsoper.com/statcalc/calculator.aspx?id=4)

In [21]:
results_first_iteration_df = pd.DataFrame(results_first_iteration)
first_iteration_profit_df = results_first_iteration_df.loc[results_first_iteration_df["description"] == "Чистий прибуток"]
first_iteration_profit_df = first_iteration_profit_df[
    [
        "experiment", 
        "price_coefficient", 
        "description", 
        "result_markers", 
        "time_modeling",
    ]
]
first_iteration_profit_df["result_markers"] = first_iteration_profit_df["result_markers"].astype("float64")
first_iteration_profit_df = first_iteration_profit_df.reset_index(drop=True)
first_iteration_profit_df["global_average"] = first_iteration_profit_df["result_markers"].mean()
first_iteration_profit_df["price_coefficient_avg_result"] = (
    first_iteration_profit_df
    .groupby(["price_coefficient"])["result_markers"]
    .transform(lambda x: x.mean())
)
first_iteration_profit_df["group_deviation"] = ((
    first_iteration_profit_df["result_markers"] - first_iteration_profit_df["price_coefficient_avg_result"])**2
)
first_iteration_profit_df["s_factual"] = (
    needed_number_of_experiments * 
    sum(
        (
            first_iteration_profit_df.groupby(["price_coefficient"])["result_markers"].mean() - 
            first_iteration_profit_df["global_average"].iloc[0]
        )**2
    )
)
first_iteration_profit_df["s_residual"] = first_iteration_profit_df["group_deviation"].sum()

num_of_groups = len(first_iteration_profit_df["price_coefficient"].unique())

d_factual = first_iteration_profit_df["s_factual"].iloc[0]
d_residual = (
    first_iteration_profit_df["s_residual"].iloc[0] / 
    ((num_of_groups * needed_number_of_experiments) - num_of_groups)
)

F = d_factual / d_residual
F_critical = 1.63654249

print(f"Sum of Squares: {d_factual}")
print(f"Degrees of Freedom: {((num_of_groups * needed_number_of_experiments) - num_of_groups)}")
print(f"Mean Square: {d_residual}")
print(f"F-actual: {F}; F-critical(alpha=0.15): {F_critical}")
print(f"Factor is {'not ' if F < F_critical else ''}significant!")

display(first_iteration_profit_df)

Sum of Squares: 2.527396837249382e+49
Degrees of Freedom: 70
Mean Square: 3.744922990530315e+47
F-actual: 67.48861975640999; F-critical(alpha=0.15): 1.63654249
Factor is significant!


experiment  price_coefficient      description                   result_markers  time_modeling                  global_average     price_coefficient_avg_result                                    group_deviation                                          s_factual                                         s_residual
0      0.00000            1.00000  Чистий прибуток                 -104569435.05700         405144 -233892548759559010779136.00000                 -122010686.30664                              304197245152941.62500 25273968372493819690654100575328421838392381145... 26214460933712208203732744398657000559287480614...
1      1.00000            1.00000  Чистий прибуток                 -119012540.68100         412741 -233892548759559010779136.00000                 -122010686.30664                                8988877192522.40430 25273968372493819690654100575328421838392381145... 26214460933712208203732744398657000559287480614...
2      2.00000            1.00000  Чистий прибуток                 -118320808.48400         413026 -233892548759559010779136.00000                 -122010686.30664                               13615198345983.57812 25273968372493819690654100575328421838392381145... 26214460933712208203732744398657000559287480614...
3      3.00000            1.00000  Чистий прибуток                 -145862279.72800         426579 -233892548759559010779136.00000                 -122010686.30664                              568898508738037.12500 25273968372493819690654100575328421838392381145... 26214460933712208203732744398657000559287480614...
4      4.00000            1.00000  Чистий прибуток                 -104639543.24600         404318 -233892548759559010779136.00000                 -122010686.30664                              301756611233094.12500 25273968372493819690654100575328421838392381145... 26214460933712208203732744398657000559287480614...
5      5.00000            1.00000  Чистий прибуток                 -139751904.42900         423779 -233892548759559010779136.00000                 -122010686.30664                              314750820465284.12500 25273968372493819690654100575328421838392381145... 26214460933712208203732744398657000559287480614...
6      6.00000            1.00000  Чистий прибуток                 -147332123.76900         428007 -233892548759559010779136.00000                 -122010686.30664                              641175195160393.00000 25273968372493819690654100575328421838392381145... 26214460933712208203732744398657000559287480614...
7      7.00000            1.00000  Чистий прибуток                 -157490882.69800         432913 -233892548759559010779136.00000                 -122010686.30664                             1258844335969735.25000 25273968372493819690654100575328421838392381145... 26214460933712208203732744398657000559287480614...
8      8.00000            1.00000  Чистий прибуток                 -117786935.76400         412675 -233892548759559010779136.00000                 -122010686.30664                               17840068646420.85938 25273968372493819690654100575328421838392381145... 26214460933712208203732744398657000559287480614...
9      9.00000            1.00000  Чистий прибуток                  -95358524.25400         397942 -233892548759559010779136.00000                 -122010686.30664                              710337742079989.37500 25273968372493819690654100575328421838392381145... 26214460933712208203732744398657000559287480614...
10    10.00000            1.00000  Чистий прибуток                  -91992571.26300         395387 -233892548759559010779136.00000                 -122010686.30664                              901087230772987.00000 25273968372493819690654100575328421838392381145... 26214460933712208203732744398657000559287480614...
11     0.00000            1.40000  Чистий прибуток                  -63652205.70600         377098 -233892548759559010779136.00000                 -101051212.53464                             13986857

In [424]:
results_first_iteration_df = pd.DataFrame(results_first_iteration)
first_iteration_profit_df = results_first_iteration_df.loc[results_first_iteration_df["description"] == "Чистий прибуток"]
first_iteration_profit_df = first_iteration_profit_df[
    [
        "experiment", 
        "price_coefficient", 
        "description", 
        "result_markers", 
        "time_modeling",
    ]
]
first_iteration_profit_df["result_markers"] = first_iteration_profit_df["result_markers"].astype("float64")
first_iteration_profit_df["rank"] = first_iteration_profit_df["result_markers"].rank(method="max", ascending=False)
first_iteration_profit_df.sort_values(by=['rank'], inplace=True)
first_iteration_profit_df = first_iteration_profit_df[first_iteration_profit_df["rank"] <= 10]
first_iteration_profit_df = first_iteration_profit_df.reset_index(drop=True)

print("Top 10 results:")
display(first_iteration_profit_df)
print("Grouped results based on price_coefficient:")
display(first_iteration_profit_df.groupby(["price_coefficient"]).size())

Top 10 results:


experiment  price_coefficient      description  result_markers  time_modeling     rank
0     0.00000            1.40000  Чистий прибуток -63652205.70600         377098  1.00000
1     3.00000            1.40000  Чистий прибуток -79959085.53100         392584  2.00000
2     9.00000            1.40000  Чистий прибуток -81146643.11300         393213  3.00000
3     7.00000            1.90000  Чистий прибуток -81557948.29900         402853  4.00000
4     2.00000            1.40000  Чистий прибуток -82237751.64100         393550  5.00000
5     2.00000            1.90000  Чистий прибуток -83518096.69900         402038  6.00000
6     5.00000            1.90000  Чистий прибуток -84158341.58800         403595  7.00000
7    10.00000            1.90000  Чистий прибуток -84575069.62700         402041  8.00000
8     4.00000            1.90000  Чистий прибуток -86198839.49200         407286  9.00000
9     1.00000            1.40000  Чистий прибуток -86699382.42700         396536 10.00000

Grouped results baseo on price_coefficient:


price_coefficient
1.40000    5
1.90000    5
dtype: int64

**Аналіз результатів**: отримали два вагомі значення коефіцієнтів, а сама `[1.4, 1.9]`. Тому надалі є доцільним обирати значення з цього діапазону.
Проведемо другу ітерацію пошуку найкращого параметру на наступному наборі значень: 
```python 
m in [1.4, 1.7, 1.9, 1.999]
```

In [425]:
results_second_iteration = []
for m in [1.4, 1.7, 1.9, 1.999]:
    for i in np.arange(needed_number_of_experiments):
        print(f"price_coefficient: {m}; experiment {i}")
        store_system = StoreSystem(m)
        for result_item in store_system.simulate(flag_display_results=False): 
            results_second_iteration.append(
                {
                    "experiment": i,
                    "price_coefficient":m,
                    **result_item
                }
            )

price_coefficient: 1.4; experiment 0.0
price_coefficient: 1.4; experiment 1.0
price_coefficient: 1.4; experiment 2.0
price_coefficient: 1.4; experiment 3.0
price_coefficient: 1.4; experiment 4.0
price_coefficient: 1.4; experiment 5.0
price_coefficient: 1.4; experiment 6.0
price_coefficient: 1.4; experiment 7.0
price_coefficient: 1.4; experiment 8.0
price_coefficient: 1.4; experiment 9.0
price_coefficient: 1.4; experiment 10.0
price_coefficient: 1.7; experiment 0.0
price_coefficient: 1.7; experiment 1.0
price_coefficient: 1.7; experiment 2.0
price_coefficient: 1.7; experiment 3.0
price_coefficient: 1.7; experiment 4.0
price_coefficient: 1.7; experiment 5.0
price_coefficient: 1.7; experiment 6.0
price_coefficient: 1.7; experiment 7.0
price_coefficient: 1.7; experiment 8.0
price_coefficient: 1.7; experiment 9.0
price_coefficient: 1.7; experiment 10.0
price_coefficient: 1.9; experiment 0.0
price_coefficient: 1.9; experiment 1.0
price_coefficient: 1.9; experiment 2.0
price_coefficient: 1.9;

_Для уникнення очікування виконання експериментів, результати були збережені. Для відновлення необхідно запустити наступну ділянку коду_

In [ ]:
results_second_iteration_path = (
    "experiments/price_coefficient_change_11_runs_2_iteration.json"
)
results_second_iteration = []
with open(results_second_iteration_path, "r") as f:
    for line in f.readlines():
        results_second_iteration.append(json.loads(line))

Аналіз результатів та перевірка значимості фактора за критерієм Фішера.
**_Помітка_**: для визначння ```F_critical``` був використаний сервіс [![Free Statistics Calculators](references/free_statistics_calculators.png "Free Statistics Calculators")](https://www.danielsoper.com/statcalc/calculator.aspx?id=4)

In [429]:
results_second_iteration_df = pd.DataFrame(results_second_iteration)
second_iteration_profit_df = results_second_iteration_df.loc[results_second_iteration_df["description"] == "Чистий прибуток"]
second_iteration_profit_df = second_iteration_profit_df[
    [
        "experiment", 
        "price_coefficient", 
        "description", 
        "result_markers", 
        "time_modeling",
    ]
]
second_iteration_profit_df["result_markers"] = second_iteration_profit_df["result_markers"].astype("float64")
second_iteration_profit_df = second_iteration_profit_df.reset_index(drop=True)
second_iteration_profit_df["global_average"] = second_iteration_profit_df["result_markers"].mean()
second_iteration_profit_df["price_coefficient_avg_result"] = (
    second_iteration_profit_df
    .groupby(["price_coefficient"])["result_markers"]
    .transform(lambda x: x.mean())
)
second_iteration_profit_df["group_deviation"] = ((
    second_iteration_profit_df["result_markers"] - second_iteration_profit_df["price_coefficient_avg_result"])**2
)
second_iteration_profit_df["s_factual"] = (
    needed_number_of_experiments * 
    sum(
        (
            second_iteration_profit_df.groupby(["price_coefficient"])["result_markers"].mean() - 
            second_iteration_profit_df["global_average"].iloc[0]
        )**2
    )
)
second_iteration_profit_df["s_residual"] = second_iteration_profit_df["group_deviation"].sum()

num_of_groups = len(second_iteration_profit_df["price_coefficient"].unique())

d_factual = second_iteration_profit_df["s_factual"].iloc[0]
d_residual = (
    second_iteration_profit_df["s_residual"].iloc[0] / 
    ((num_of_groups * needed_number_of_experiments) - num_of_groups)
)

F = d_factual / d_residual
F_critical = 1.87109477

print(f"Sum of Squares: {d_factual}")
print(f"Degrees of Freedom: {((num_of_groups * needed_number_of_experiments) - num_of_groups)}")
print(f"Mean Square: {d_residual}")
print(f"F-actual: {F}; F-critical(alpha=0.15): {F_critical}")
print(f"Factor is {'not ' if F < F_critical else ''}significant!")

display(second_iteration_profit_df)

Sum of Squares: 1651063537097832.8
Degrees of Freedom: 40.0
Mean Square: 312621868988156.56
F-actual: 5.281343696273475; F-critical(alpha=0.15): 1.87109477
Factor is significant!


experiment  price_coefficient      description   result_markers  time_modeling  global_average  price_coefficient_avg_result        group_deviation              s_factual              s_residual
0      0.00000            1.40000  Чистий прибуток  -82225737.20100         392438 -96071651.54364              -103358453.43536  446591695442136.12500 1651063537097832.75000 12504874759526262.00000
1      1.00000            1.40000  Чистий прибуток  -84790183.79200         394911 -96071651.54364              -103358453.43536  344780637548659.68750 1651063537097832.75000 12504874759526262.00000
2      2.00000            1.40000  Чистий прибуток  -95402846.48700         405065 -96071651.54364              -103358453.43536   63291681916851.70312 1651063537097832.75000 12504874759526262.00000
3      3.00000            1.40000  Чистий прибуток -121947377.32700         422119 -96071651.54364              -103358453.43536  345548091449049.50000 1651063537097832.75000 12504874759526262.00000
4      4.00000            1.40000  Чистий прибуток  -89411393.87800         399199 -96071651.54364              -103358453.43536  194520470296648.15625 1651063537097832.75000 12504874759526262.00000
5      5.00000            1.40000  Чистий прибуток -135417563.19100         428013 -96071651.54364              -103358453.43536 1027786518323939.62500 1651063537097832.75000 12504874759526262.00000
6      6.00000            1.40000  Чистий прибуток -112093053.81300         416041 -96071651.54364              -103358453.43536   76293243757005.20312 1651063537097832.75000 12504874759526262.00000
7      7.00000            1.40000  Чистий прибуток  -69067164.36400         382583 -96071651.54364              -103358453.43536 1175892506175823.50000 1651063537097832.75000 12504874759526262.00000
8      8.00000            1.40000  Чистий прибуток -114281348.21000         415939 -96071651.54364              -103358453.43536  119309630257778.25000 1651063537097832.75000 12504874759526262.00000
9      9.00000            1.40000  Чистий прибуток -110212627.49600         412628 -96071651.54364              -103358453.43536   46979702053500.48438 1651063537097832.75000 12504874759526262.00000
10    10.00000            1.40000  Чистий прибуток -122093692.03000         421275 -96071651.54364              -103358453.43536  351009165197952.00000 1651063537097832.75000 12504874759526262.00000
11     0.00000            1.70000  Чистий прибуток -100812860.99400         413173 -96071651.54364              -100895638.28127       6852079288.23313 1651063537097832.75000 12504874759526262.00000
12     1.00000            1.70000  Чистий прибуток  -89259700.91700         405171 -96071651.54364              -100895638.28127  135395038345278.51562 1651063537097832.75000 12504874759526262.00000
13     2.00000            1.70000  Чистий прибуток -111683437.13900         420454 -96071651.54364              -100895638.28127  116376604194781.56250 1651063537097832.75000 12504874759526262.00000
14     3.00000            1.70000  Чистий прибуток  -87079012.06400         402031 -96071651.54364              -100895638.28127  190899160027828.53125 1651063537097832.75000 12504874759526262.00000
15     4.00000            1.70000  Чистий прибуток  -64258520.18000         383031 -96071651.54364              -100895638.28127 1342278422766606.50000 1651063537097832.75000 12504874759526262.00000
16     5.00000            1.70000  Чистий прибуток -127316638.61200         429404 -96071651.54364              -100895638.28127  698069258476290.25000 1651063537097832.75000 12504874759526262.00000
17     6.00000            1.70000  Чистий прибуток -103990704.61800         415601 -96071651.54364              -100895638.28127    9579435628742.31055 1651063537097832.75000 12504874759526262.00000
18     7.00000            1.70000  Чистий прибуток -111670628.34300         420321 -96071651.54364              -100895638.28127  116100410830321.12500 1651063537097832.75000 12504874759526262.00000
19     8.00000          

In [431]:
results_second_iteration_df = pd.DataFrame(results_second_iteration)
second_iteration_profit_df = results_second_iteration_df.loc[results_second_iteration_df["description"] == "Чистий прибуток"]
second_iteration_profit_df = second_iteration_profit_df[
    [
        "experiment", 
        "price_coefficient", 
        "description", 
        "result_markers", 
        "time_modeling",
    ]
]
second_iteration_profit_df["result_markers"] = second_iteration_profit_df["result_markers"].astype("float64")
second_iteration_profit_df["rank"] = second_iteration_profit_df["result_markers"].rank(method="max", ascending=False)
second_iteration_profit_df.sort_values(by=['rank'], inplace=True)
second_iteration_profit_df = second_iteration_profit_df[second_iteration_profit_df["rank"] <= 10]
second_iteration_profit_df = second_iteration_profit_df.reset_index(drop=True)

print("Top 10 results:")
display(second_iteration_profit_df)
print("Grouped results based on price_coefficient:")
display(second_iteration_profit_df.groupby(["price_coefficient"]).size())

Top 10 results:


experiment  price_coefficient      description  result_markers  time_modeling     rank
0     4.00000            1.99900  Чистий прибуток -58796955.06200         385532  1.00000
1     4.00000            1.70000  Чистий прибуток -64258520.18000         383031  2.00000
2     7.00000            1.40000  Чистий прибуток -69067164.36400         382583  3.00000
3     3.00000            1.90000  Чистий прибуток -69194708.53100         390688  4.00000
4     6.00000            1.90000  Чистий прибуток -72074814.29300         395600  5.00000
5     5.00000            1.90000  Чистий прибуток -76703217.35300         399665  6.00000
6     0.00000            1.99900  Чистий прибуток -77100715.30700         403401  7.00000
7     6.00000            1.99900  Чистий прибуток -78296677.13300         400837  8.00000
8    10.00000            1.99900  Чистий прибуток -79668683.51400         402402  9.00000
9     0.00000            1.90000  Чистий прибуток -80120183.06400         399173 10.00000

Grouped results baseo on price_coefficient:


price_coefficient
1.40000    1
1.70000    1
1.90000    4
1.99900    4
dtype: int64

**Аналіз результатів**: значення коефіцієнта `1.999` займає першу позацію та зустрічається в топ-10 чотири рази, тому можна дійти до висновку, що це і є оптимальним значенням фактора, що є цілком логічним, оскільки саме це значення є граничним при переході різниць ймовірностей. Значення цільової функції в найкращому випадку: `-58796955.06200`. Тобто, робимо висковки, що підприємство **_не є прибутковим_**.

#### Додаткова перевірка
Для того, щоб переконатися, що система моделюється правильно, був проведений запуск відповідно за умов варіанту 1. 
За результати помітно, що прибуток є більшим за нуля. Це зумовлене тим, що час надходження покупців та розподіл ймовірностей відповідно до ціни товару є кращим ніж у варіанті 2.

In [52]:
_system = StoreSystem(1.999)
_system.simulate()


# LAMBDA = 20
# k = 0.01
# s = 1000
# N = 600

# p1 = {
#     (0, 5): 0.3,
#     (6, 10): 0.3,
#     (11, 20): 0.2,
#     (21, 50): 0.1,
#     (51, sys.maxsize): 0.1,
# }
# p2 = {
#     (0, 1.4999): 0.45,
#     (1.5, 1.9999): 0.4,
#     (2.0, 2.9999): 0.06,
#     (3.0, 3.9999): 0.04,
#     (4.0, 5.9999): 0.03,
#     (6.0, 9.9999): 0.02,
#     (10.0, sys.maxsize): 0,
# }

# A = 30
# B = 5

# m = 1.999
# M = 1000


Modeling time spent: 103447


description   avg_markers  max_markers  min_markers  result_markers
0                            Потік покупців  5.801028e-03            1            0               0
1                         Покупець надійшов  5.801028e-03            1            0               0
2                            Черга покупців  9.675905e-03            2            0               0
3   Кількість покупців, що покинули мазагин  1.397326e+03         4282            0            4282
4                              Запас товару  4.004802e+02          600            0               0
5                                 Продавець  8.321457e-01            1            0               1
6           Кількість обслугованих покупців  1.993519e+02          600            0             600
7                                  Прибуток  1.289389e+07    466719000            0       466719000
8                                      Богр  3.028773e+08    600000000          588             776
9                            Погашений борг  3.856106e+08    732681000            0       732681000
10               Затримка процентної ставки  8.436562e-05            1            0               0
11           Процентна ставка готова до дії  8.436562e-05            1            0               0

[{'description': 'Потік покупців',
  'avg_markers': 0.005801027834714672,
  'max_markers': 1,
  'min_markers': 0,
  'result_markers': 0},
 {'description': 'Покупець надійшов',
  'avg_markers': 0.005801027834714672,
  'max_markers': 1,
  'min_markers': 0,
  'result_markers': 0},
 {'description': 'Черга покупців',
  'avg_markers': 0.009675905296616463,
  'max_markers': 2,
  'min_markers': 0,
  'result_markers': 0},
 {'description': 'Кількість покупців, що покинули мазагин',
  'avg_markers': 1397.326266821139,
  'max_markers': 4282,
  'min_markers': 0,
  'result_markers': 4282},
 {'description': 'Запас товару',
  'avg_markers': 400.48020675519115,
  'max_markers': 600,
  'min_markers': 0,
  'result_markers': 0},
 {'description': 'Продавець',
  'avg_markers': 0.8321456792323917,
  'max_markers': 1,
  'min_markers': 0,
  'result_markers': 1},
 {'description': 'Кількість обслугованих покупців',
  'avg_markers': 199.35193892404124,
  'max_markers': 600,
  'min_markers': 0,
  'result_markers':